fc/code_generation.py

```diff
@@ -132,6 +133,13 @@ def __init__(self, symbol_function=None, derivative_function=None):
         # Deal with _exp function introduced to avoid simplification
         self._function_names['_exp'] = 'math.exp'
 
+    def doprint(self, expr):
```

Comment:
Admittedly I don't understand the sympy printer design, but this seems an expensive way to resolve the issue! Is the replace recursive throughout the whole expression, or just at a single level? In any case, shouldn't we be able to define a `_print_DataInterpolation` method in this class which will be used, according to https://docs.sympy.org/latest/modules/printing.html?

Reply:
the problem is that the super().do_print does a replace optimisation that causes wierd behaviour **BEFORE** you would get to any `_print_DataInterpolation`. This call causes DataInterpolation.__init__ to be called with just a cellmlmanip.model.Variable as parameter and I can't figure out what to do in that case. If I just spit it back or make a copy I get a strange error, which I don't get with either the current work around or if I knowck out the replace optimisation in the printer.

It probably would be possible to avoid the replace, but that doesn't solve the real issue in that doing a replace (even on things that aren't even in the expression) causes this srang behaviour.

```
model.c:6748:15: warning: '__pyx_f_5model_lookup_data_table_2' defined but not used [-Wunused-function]
 static double __pyx_f_5model_lookup_data_table_2(PyObject *__pyx_v_var_protocol__table_key_converted) {
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
model.c:6609:15: warning: '__pyx_f_5model_lookup_data_table_0' defined but not used [-Wunused-function]
 static double __pyx_f_5model_lookup_data_table_0(PyObject *__pyx_v_var_time_converted) {
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
model.c:6470:15: warning: '__pyx_f_5model_lookup_data_table_1' defined but not used [-Wunused-function]
 static double __pyx_f_5model_lookup_data_table_1(PyObject *__pyx_v_var_protocol__derived2_converted) {
```

In fact without that replace the whole `_print_DataInterpolation` isn't even necessary (it's was previously solved in the `variable_name` method in code_generation.py)

Reply:
The `Variable` argument is probably because of the line `obj._args = [index_variable]` in `__new__`, so something is trying to create a new `DataInterpolation` just from its `_args`. Adding more things to `_args` might well break some other part of sympy's machinery, I don't know!

Looking at the latest code for `Symbol` etc., maybe we need to define at least `__getnewargs__`. Or actually from [`Basic`](https://github.com/sympy/sympy/blob/9e8f62e059d83178c1d8a1e19acac5473bdbf1c1/sympy/core/basic.py#L117) maybe define our own `copy` to do the right thing? Or our own [`func`](https://github.com/sympy/sympy/blob/9e8f62e059d83178c1d8a1e19acac5473bdbf1c1/sympy/core/basic.py#L682).

Reply:
Yes adding more things to args breaks other things I did try that. 
I did try define a copy/deepcopy but it doesn't get called.

I'll look into basic/func.